### PR TITLE
[FIX JENKINS-34855] Create a FileChannelWriter and use it for AtomicFileWriter "enforced" flushing

### PR DESCRIPTION
**Pull request by batmat**
See [JENKINS-34855](https://issues.jenkins-ci.org/browse/JENKINS-34855).

Followup of #2548 & #3165

After a lot of reading and some testing, the only reliable ways to force a **local** device sync from Java code appear to be either by using `FileChannel.force(boolean)` or `FileDescriptor.sync()`.

So even if the issues around this have generally been seen more when using remote FS, I think we can hope (?) that making sure we actually flush to local disk could at least make the probability for that issue (i.e. zero-length files seen in the wild) to happen overall (far?) less _likely_.

This change rewires the current internal `core` writer in `AtomicFileWriter` to delegate to a `FileChannel`, giving then us access to `FileChannel.force()`.

### Proposed changelog entries

* `Internal:` [JENKINS-34855](https://issues.jenkins-ci.org/browse/JENKINS-34855) `AtomicFileWriter` now uses a `FileChannel` internally and forces a disk sync when `commit`ting.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

#### ToDos after merge

* [ ] Update https://wiki.jenkins.io/display/JENKINS/Features+controlled+by+system+properties with the new sysprop
* [ ] File an improvement followup from James' comment: https://github.com/jenkinsci/jenkins/pull/3170#pullrequestreview-80799082
 
### Desired reviewers

@reviewbybees esp. @jtnord

### CODE REVIEW

File: test/src/test/java/hudson/util/AtomicFileWriterPerfTest.java [hunk 1]
```diff
@@ -24,7 +24,7 @@
      * <strong>really</strong> bad performance regressions.
      */
     @Issue("JENKINS-34855")
-    @Test(timeout = 30 * 1000L)
+    @Test(timeout = 100 * 1000L)
```
Comment (batmat) [inline test/src/test/java/hudson/util/AtomicFileWriterPerfTest.java]: FTR:

image

File: test/src/test/java/hudson/util/AtomicFileWriterPerfTest.java [hunk 2]
```diff
@@ -24,7 +24,7 @@
      * <strong>really</strong> bad performance regressions.
      */
     @Issue("JENKINS-34855")
-    @Test(timeout = 30 * 1000L)
+    @Test(timeout = 80 * 1000L)
```
Comment (amuniz) [inline test/src/test/java/hudson/util/AtomicFileWriterPerfTest.java]: Why? Has this worsened performance by ~2x?

**Review by amuniz — COMMENTED** [review]
Did you see some flakiness after the fix in #2548? I wonder if a "corner case" is worth adding this complexity... I've seen the "atomic not being atomic" issue on network file systems only, never locally.

File: test/src/test/java/hudson/util/AtomicFileWriterPerfTest.java (hunk 2 above)
Comment (batmat) [reply test/src/test/java/hudson/util/AtomicFileWriterPerfTest.java]: Worse than that, around ~30+ for that very case. I just reduced to 50 seconds, since it actually takes ~35 seconds in CI.
 
But as explained in the [commit log](https://github.com/jenkinsci/jenkins/pull/3170/commits/f555cb6e7a0d4199a4417841e7b42e7f61786c22), I still think the overall change is an improvement. Before, we were quicker, but unsafe, is the whole point of this... :-|

**Comment by batmat** [comment]
>"corner case"

This fix is actually more like the _main_ one, #2548 is only about atomic rename. But it could still happen we rename something that has still bits not flushed in the myriads of layers of FS caching and networking.

The whole issue may indeed be a "corner case", granted, but is a dramatic one when it happens as you know ("zero-length"ed config files can just take down the whole instance). 

I manually built and ran Jenkins on my machine with this change, and couldn't see any noticeable degradation. And that makes sense: for instance, `AtomicFileWriterPerfTest` does indeed go from 1 second to 35 seconds. 

Taken out of context, it can look as just a horrible perf regression.

But now, if one looks at the test (and I wrote it only a few days ago to _monitor_ this), this is doing **1000** `Jenkins.save()`. So, before, that means a `Jenkins.save()` now takes (on the CI setup) ~35 milliseconds, instead of previously ~1 millisecond. 

_35x_ **is** a huge variation, but the absolute value IMO tells us that that shouldn't be an actual issue, knowing also we have between those two added a far better effort to make sure important things are **actually** written to disk.

File: core/src/main/java/hudson/util/FileChannelWriter.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            channel.force(true);
+        } else {
+            LOGGER.finest("Force disabled on flush(), no-op");
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        if(channel.isOpen()) {
+            channel.force(true);
```
Comment (jtnord) [inline core/src/main/java/hudson/util/FileChannelWriter.java]: for the case where the file is being thrown away (abort is called rather than commit) we could get away with not calling `force`.  (that is the ATF.commit could call a force method which does a flush, force, and then subsequently call close)

**Review by jtnord — APPROVED** [review]

File: core/src/main/java/hudson/util/FileChannelWriter.java (hunk 3 above)
Comment (batmat) [reply core/src/main/java/hudson/util/FileChannelWriter.java]: Will do as a followup.

**Comment by jtnord** [comment]
@amuniz 
> I've seen the "atomic not being atomic" issue on network file systems only, never locally.

I've seen it locally, depends on your file system of choice, especially if some flush metadata before data (or independantly)...  https://www.spinics.net/lists/xfs/msg09911.html

**Comment by batmat** [comment]
@jenkinsci/code-reviewers for possibly more eyes or testing.

**Review by dwnusbaum — DISMISSED** [review]
🐝

**Review by svanoort — CHANGES_REQUESTED** [review]
While I think this does ensure commit() really *does*, I feel due to the performance discrepancy, the enforced sync-to-disk should be controlled by a system property, with default set to off. 

That way for cases where we need ABSOLUTE integrity, we can set the property on (covering original issues), but we don't wreck the majority of masters where that is not the case. 

I can tell you that because this behavior AFAICT is on for all XmlFiles, and Pipeline hits those heavily until JENKINS-47170 lands, the second this lands no master will be able to run Pipeline at an acceptable rate.

Or is this already covered by something else in the other changes that I've missed?

Edit: opt-in enforced flushing for the most critical config files (Jenkins core configs + Job config) would also be acceptable IMO, esp if a system property allows you to disable autosync if we encounter edge cases.  It's more the files related to builds or that are touched often where it's a major problem.

**Comment by jtnord** [comment]
> That way for cases where we need ABSOLUTE integrity, we can set the property on (covering original issues), but we don't wreck the majority of masters where that is not the case.

@svanoort you can not control requiring integrity vs not - it is just not an option.

most people may thing oh yeah - but then Jenkins starts with an empty / missing config.xml and everything goes to pot - and it is Jenkins fault.

Code that does not need to care about integrity has no business using `AtomicFileWriter` to begin with - as the sole purpose of that was to ensure integrity of the system. 

If you ask people to set a flag once they have dataloss you will be screamed at "WHY WAS THAT NOT THE <insert expletive> DEFAULT FOR <insert deity> SAKE"

If you want a flag then the flag then this should be of the form `"I_DO_NOT_CARE_ABOUT_SYSTEM_INTEGRITY_GIVE_ME_PERFORMANCE"` and have a value of `"I_ACCEPT_THE_RISK"` before we throw integrety out of the window.

That is you can OPT out of integrity but you never have to opt into it when it is the expected behaviour.

**Comment by jglick** [comment]
I would tend to agree. Jenkins should be safe and correct by default. If there is a demonstrable I/O performance issue under heavy load conditions involving Pipeline, well we can always have a system property to opt out of integrity guarantees, but better would be for the Pipeline `FlowNode` storage engine to selectively avoid `AtomicFileWriter` under conditions where it knows loss of a file is tolerable. (The same might be true of `build.xml` for `Run`s generally—if these get truncated, well, life moves on.)

**Comment by batmat** [comment]
@jglick @svanoort what plugin(s) you think already have tests (or could have?) that would assess the impact it would have (running with -Djenkins.version=2.94-SNAPSHOT)?

As explained above, though the impact is big on proportion, especially on ridiculously slow disks, we're still talking about milliseconds in absolute values. So for instance if flownodes are saved in a single build.xml (?), I tend to think the impact shouldn't be gigantic?
I disabled the SYNC on flush() for this purpose (because XmlFile transitively calls flush for every single writes or so), or do we call `AtomicFileWriter.commît()` potentially like dozens or hundreds of times per second from Pipeline code?

**Comment by svanoort** [comment]
@batmat I ought to be able to test it against Pipeline with Hydra, AKA The Scalability Lab soonishly -- there's a setting I need to benchmark anyway that turns *OFF* all the AtomicWriter behavior for Pipeline, so that'll be an easy approach to before/after comparison. 

The problem with FlowNodes is that they're (currently) saved by default in a whole bunch of tiny XML files, and those small overheads already add up in a big way for them.  Now, I have a mode that does away with that nonsense but people have to know that this is the problem and they need to select it.

I suspect that's not the only place where this overhead will pose a problem or where there's XML saved often.

In general, I'm not a fan of changes that introduce huge, non-optional performance regressions to protect against a rare failure mode (tied to operating-system level failures, here). To a moderate-to-large scale users, these performance changes are more likely to bring down a master than an edge case like this.  I remind you all here about the ongoing pain we've had with user lookup performance.  "Just" a performance regression there has claimed the lives of many production masters at large companies, c.f. [JENKINS-35493](https://issues.jenkins-ci.org/browse/JENKINS-35493) between this and its recent kin (c.f. [JENKINS-47429](https://issues.jenkins-ci.org/browse/JENKINS-47429).

> Code that does not need to care about integrity has no business using AtomicFileWriter to begin with - as the sole purpose of that was to ensure integrity of the system.

@jtnord The problem is that `XmlFile.write(Object)` is what *many things* use by default for persisting XStream content. 

I think @batmat did a really clever fix here, but we need to ensure that the fix is only applied where it's critical, i.e top-level Jenkins Configs or Build configs, or that at least people can turn it off.

I believe the phrase here is "First Do No Harm"?

**Comment by stephenc** [comment]
> "First Do No Harm"

Which is exactly picking integrity over performance (as the default... it’s fine to allow users to opt-in to picking performance over integrity)

**Comment by svanoort** [comment]
@stephenc As we all know, most users are **not** going to be happy if a core update breaks their master.  For them there's not much difference between "completely broken" and "running so slowly I can't tell it's actually working."  I'm saying that this change no only *can* cause this, it *probably will.* 

As @batmat said:

> do we call AtomicFileWriter.commît() potentially like dozens or hundreds of times per second from Pipeline code?

[Yes](https://github.com/jenkinsci/workflow-support-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/support/storage/SimpleXStreamFlowNodeStorage.java#L111), unfortunately [we](https://github.com/jenkinsci/workflow-support-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/support/storage/SimpleXStreamFlowNodeStorage.java#L93) do.  

Dozens to hundreds of times per second **is actually a very good estimate for the rate at which WorkflowRun and FlowNodes are serialized on busier masters** -- ignoring the other XML files for users, etc.  Currently a lot of these writes appear to be hitting OS write caching, so multiple writes to the same file are coalesced (greatly improving performance) -- the result is 4-5 physical IO per FlowNode rather than the 8-10 we'd see otherwise.

File: core/src/main/java/hudson/util/AtomicFileWriter.java [hunk 4]
```diff
@@ -116,7 +117,7 @@ public AtomicFileWriter(@Nonnull Path destinationPath, @Nonnull Charset charset)
             throw new IOException("Failed to create a temporary file in "+ dir,e);
         }
 
-        core = Files.newBufferedWriter(tmpPath, charset);
+        core = new FileChannelWriter(tmpPath, charset, false, StandardOpenOption.WRITE);
```
Comment (jglick) [inline core/src/main/java/hudson/util/AtomicFileWriter.java]: Would it helpful to just write to an in-memory buffer? These files are typically smallish, after all, so we do not care about heap usage much, and then we do not care at all who is `flush`ing when. If and when `commit` is called, write the entire buffer to the temp file (if you are lucky this will be a single block), make sure it is synched, then do the atomic rename.

I also think we should add a flag to `XmlFile` allowing callers to opt out of expensive consistency guarantees, with the default being strict mode. `Run.getDataFile` and `SimpleXStreamFlowNodeStorage.getNodeFile` could then use the lax mode, accepting the risk of corrupted data in exchange for better performance. You can even guide callers to make this decision explicitly by (say) deprecating existing `XmlFile` constructors in favor of ones taking a `boolean` flag.

Comment (jtnord) [reply core/src/main/java/hudson/util/AtomicFileWriter.java]: > Would it helpful to just write to an in-memory buffer? These files are typically smallish, after all, so we do not care about heap usage much, and then we do not care at all who is flushing when. If and when commit is called, write the entire buffer to the temp file (if you are lucky this will be a single block), make sure it is synched, then do the atomic rename.

sounds a little like an over eager optimisation to me.
When writing bytes to a file (as we do here without the `SYNC` / `DSYNC`  flag) the OS will generally buffer them (almost all OSes have a disk cache buffer) , and they are not written directly to disk.  
The only sync call is the one in commit in the case that `false` is passed (which it is) for AtomicFW you would always want false (as until commit is called it can be thrown away), for any direct uses of the channel writer you may want to know writes have been flushed to disk before returning.

**Comment by svanoort** [comment]
So I have some not-good benchmarks to back up my point.  On the left we see a load test environment running 4 concurrent builds of a series of pipeline Echo steps plus some stages -- running builds continuously one after the other.  On the right we see the same workload but after swapping the Jenkins core from 2.60.3 to a build of this PR. 

screen shot 2017-12-11 at 6 51 07 pm

The top blue line is build times for a run - it goes from 1.8s to ~28s.   Then the build time steps down a bit when I disable atomic write in a few easy places (by no means a comprehensive solution though).

Even more glaring, if I look at how many flownodes (records of pipeline step execution) per minute we can run:

screen shot 2017-12-11 at 6 50 55 pm


It drops from ~30k/min to 2k/min, and then 2.5k when some atomic writes are disabled.  Not good!

Methodology note: this is set to have IO characteristics on the same order as EBS - 3000 IOPS permitted (separate read and write limits), 200 MB/s read and write throughputs.

**Comment by jtnord** [comment]
@svanoort thanks for the details - but all that prooves to me is that pipeline should not be using `AtomicFileWriter` if it does not care about integrity or that the file exists after a restart (ie it can tollerate a failure).

It does not say to me that this needs to be opt in by default.

> Methodology note: this is set to have IO characteristics on the same order as EBS with an SSD-backed GP2 volume and bursting -- 3000 IOPS permitted (separate read and write limits), 200 MB/s read and write throughputs.

OT: Can you do this on an actual EBS volume to show that your flush characteristics match?

**Comment by svanoort** [comment]
@jtnord Pipeline has legit reasons to avoid data loss though -- loss of AtomicFileWriter could impact the ability to resume or cleanly abort a critical production deploy, for example.  My solution is to give people the choice of how they want to handle this. 

Pipeline is also not the only plugin doing a decent amount of XML writes -- @abayer has mentioned that some of the other project types (either matrix or maven project IIRC) and JobConfigHistory do this. 

For this reason can we agree it's needed to put some escape valves here, so users don't experience "master suddenly slows down a lot" as the default?  At the very least, an opt-out option, or make the API opt-in for XML files (here we can auto opt-in most of the normal types). 

As far as infra: I haven't spun any infra for this yet -- it's on the roadmap but will take a while, since I have to tackle some Docker Swarm pain points (really just Swarm bugs/arbitrary limits) and ensure the whole thing is properly secured.   You can safely assume that the performance characteristics of this test environment are likely better than actual EBS -- but I'm not going to go into the details, because this isn't intended to be a scientifically precise benchmark, just a rough assessment of the impact this change has (huge).

**Comment by abayer** [comment]
fwiw, the real fun with JobConfigHistory in this regard is when it goes bonkers with a Maven job and records each module individually for each build of the job. So there you get a pretty nasty spam of XML files. Otherwise, Pipeline would write the most XML files, followed by Maven jobs, and then Matrix jobs (depending on how many configurations they've got, of course). Pipeline is head and shoulders above the others in potential number of XML files, though.

**Comment by jtnord** [comment]
> At the very least, an opt-out option

I've already said that an opt-out is OK.

**Comment by jtnord** [comment]
> fwiw, the real fun with JobConfigHistory in this regard is when it goes bonkers with a Maven job and records each module individually for each build of the job

The real fun is it does all of its work whilst holding and assuming that this is indeed safe to do so - eventually causing deadlocks.  The chance of which that you hit one of these deadlocks will likely be increased by this code as writing the files will take longer giving a greater window of opportunity., but well that's an existing issue :-/

**Comment by batmat** [comment]
So I just pushed https://github.com/jenkinsci/jenkins/pull/3170/commits/d692b094f06b89d841ea7746d92cefbc9732ec42 which adds two flags:
* a runtime one that be used programmatically for specific cases. I've deprecated the corresponding constructor to discourage people from using it lightly
* a global one, using a sysprop, to be used in case of production emergency `hudson.util.AtomicFileWriter.LOSE_DATA_INTEGRITY_FOR_PERFORMANCE`.

@svanoort care to have another look? Thanks a lot for the efforts and the deep review you did here!

**Review by jtnord — APPROVED** [review]

**Comment by batmat** [comment]
Updated as https://github.com/jenkinsci/jenkins/pull/3170/commits/3a5189ebc861832ce5189f1104fd0beff2e07233 flush & close flags were reversed.

File: core/src/main/java/hudson/util/FileChannelWriter.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        channel.write(byteBuffer);
+    }
+
+    @Override
+    public void flush() throws IOException {
+        if (forceOnFlush) {
+            LOGGER.finest("Flush is forced");
+            channel.force(true);
+        } else {
+            LOGGER.finest("Force disabled on flush(), no-op");
```
Comment (olivergondza) [inline core/src/main/java/hudson/util/FileChannelWriter.java]: Should we not do *some* flush here? Otherwise turning the flag off would be result in behaviour more broken that it was before.

Comment (batmat) [reply core/src/main/java/hudson/util/FileChannelWriter.java]: @olivergondza good question.

TL;DR: `FileChannel.write()` already achieves the equivalent of `BufferedWriter.write() + BufferedWriter.flush()`.

`FileChannel`, contrary to the previous `Files.newBufferedWriter` has no Java-level buffering. So `FileChannel.write()` writes to disk, well actually sends data to the OS/disk cache depending on the system config. 

So, flushing does not make as much sense here, or at least not the same meaning as when one calls `BufferedWriter.flush()`. And BTW, even `BufferedWriter.flush()` does not actually force a full flush to disk, it's more like a cache from JVM cache/RAM to the next level, no guaranty things end up on the disk. *That* is the whole point of this change. 

`FileChannel.force()` **actually** guarantees everything is flushed to the disk. `BufferedWriter.write() + BufferedWriter.flush()` does not.

Comment (olivergondza) [reply core/src/main/java/hudson/util/FileChannelWriter.java]: I see. So, if I understand correctly, the data should be send to disk more eagerly than before (avoiding Java level caching) even when turned off. Not that I would object against the change, I am merely pointing out there is no way for users to switch to the old behaviour we know was working fine for ages in case this would not work as expected.

Comment (jtnord) [reply core/src/main/java/hudson/util/FileChannelWriter.java]: > switch to the old behaviour we know was working fine for ages in case this would not work as expected.

`forceoflush` defaults to `false` from `AtomicFileWriter` - so we are not pushing to disk on a flush thus it should behave very similar (the data has gone from Java to the OS) - although possibly via different underlying native calls.

By default from `AtomicFileWriter` we are only making sure the data is on disk when the Writer is `close()`d  and if you want to simulate the old broken behaviour you can: see [here](https://github.com/jenkinsci/jenkins/pull/3170/files#diff-2fecb1bc52ed9dba9824e03ab7a24553R57) and [here](https://github.com/jenkinsci/jenkins/pull/3170/files#diff-2fecb1bc52ed9dba9824e03ab7a24553R146)

File: core/src/main/java/hudson/util/AtomicFileWriter.java [hunk 6]
```diff
@@ -51,6 +54,15 @@
 
     private static final Logger LOGGER = Logger.getLogger(AtomicFileWriter.class.getName());
 
+    private static /* final */ boolean LOSE_INTEGRITY = SystemProperties.getBoolean(
+            AtomicFileWriter.class.getName() + ".LOSE_DATA_INTEGRITY_FOR_PERFORMANCE");
```
Comment (svanoort) [inline core/src/main/java/hudson/util/AtomicFileWriter.java]: This name feels really overly dramatic.   Could we call it "DISABLE_FORCED_FLUSH" instead?

File: core/src/main/java/hudson/util/AtomicFileWriter.java [hunk 7]
```diff
@@ -51,6 +54,15 @@
...
+
+    static {
+        if (LOSE_INTEGRITY) {
+            LOGGER.log(Level.SEVERE, "LOSE_DATA_INTEGRITY_FOR_PERFORMANCE flag used, YOU RISK LOSING DATA.");
```
Comment (svanoort) [inline core/src/main/java/hudson/util/AtomicFileWriter.java]: Again, overly dramatic.  I think we really only need to warn people about this case happening when the OS or storage is yanked rather than cleanly shut down.

File: core/src/main/java/hudson/util/AtomicFileWriter.java [hunk 8]
```diff
@@ -116,7 +143,15 @@ public AtomicFileWriter(@Nonnull Path destinationPath, @Nonnull Charset charset)
             throw new IOException("Failed to create a temporary file in "+ dir,e);
         }
 
-        core = Files.newBufferedWriter(tmpPath, charset);
+        if (LOSE_INTEGRITY) {
+            // We should log it in WARN, but as this code is called often, this would create huge logs in prod system
+            // Hence why a similar log is logged once above in a static block.
+            LOGGER.log(Level.FINE, "WARNING: YOU SET A FLAG THAT COULD LEAD TO DATA LOSS.");
```
Comment (svanoort) [inline core/src/main/java/hudson/util/AtomicFileWriter.java]: Seriously, we don't need to warn on every write.  The data loss happens when you're yanking storage/OS out from under the software -- most applications aren't expected to behave in that scenario.

Comment (svanoort) [reply core/src/main/java/hudson/util/AtomicFileWriter.java]: We should be issuing a more specific warning, something like "virtualization or storage failures can cause dataloss if this is enabled."  My concern is that in settings where users *do* have reliable storage and are bitten by the IO performance drop they will still be too scared to use this even if we explicitly tell 'em to.

File: core/src/main/java/hudson/util/AtomicFileWriter.java (hunk 6 above)
Comment (svanoort) [reply core/src/main/java/hudson/util/AtomicFileWriter.java]: Basic rule: an option should tell people something about what it does -- this isn't a very descriptive name IMO and isn't really accurate either since it's not globally destroying data integrity, it's modifying one specific behavior wrt the OS caching.

Comment (batmat) [reply core/src/main/java/hudson/util/AtomicFileWriter.java]: I agree to adjust this, but I think our current behaviour is also not 100% safe enough by default, which it should probably be. I could have put something like `REDUCE_LIKELINESS_TO_LOSE_DATA_WHEN_THINGS_GO_WRONG` and that would indeed have been closer to the reality, but well :-).

`DISABLE_FORCED_FLUSH` feels a bit not dramatic enough. My thought and understanding of this is that we really would like to not have it disabled apart from emergencies and very special cases, so let's not encourage it.

But OK, let's rename the option, and still leave a (less dramatic too, ok) message on class loading.

**Comment by batmat** [comment]
@svanoort drama adjusted ;-)

**Review by olivergondza — APPROVED** [review]

**Review by svanoort — APPROVED** [review]
Nice work here @batmat I think with the clearer warning & flag name we are in good shape.

**Comment by oleg-nenashev** [comment]
Please give me some time to review it guys. I think it's generally fine to go into the next LTS

**Comment by batmat** [comment]
@reviewbybees done.

**Comment by oleg-nenashev** [comment]
Won't have time for it this week, sorry.
Feel free to go forward if you feel it's ready

**Comment by batmat** [comment]
Thank you very much for trying @oleg-nenashev.

Yes, I think it's ready. It's been reviewed pretty thoroughly by many people, and I've myself tested it quite a lot. I would like it to land sooner than later in weekly so that we can possibly detect more (hopefully not) unforeseen impacts in the wild, and fix them all so that JENKINS-34855 can be backported and land into next LTS.

So I will move forward and merge this. Please ping me if this is deemed to have caused any issues once released. 

Thank you everyone for the deep reviews.

**Review by oleg-nenashev — APPROVED** [review]
I see no obvious issues in the code.
